packages/itwinui-react/src/core/ComboBox/ComboBoxMenu.tsx

```diff
@@ -49,18 +58,42 @@ const VirtualizedComboBoxMenu = (props: React.ComponentProps<'div'>) => {
     );
   }, [focusedIndex, menuRef]);
 
-  const { outerProps, innerProps, visibleChildren } = useVirtualization({
-    // 'Fool' VirtualScroll by passing length 1
-    // whenever there is no elements, to show empty state message
-    itemsLength: filteredOptions.length || 1,
-    itemRenderer: virtualItemRenderer,
-    scrollToIndex: focusedVisibleIndex,
-  });
+  useLayoutEffect(() => {
+    if (focusedVisibleIndex) {
+      virtualizer.scrollToIndex(focusedVisibleIndex, { align: 'start' });
+    }
+  }, [focusedVisibleIndex, virtualizer]);
+
+  const innerProps = {
+    style: { willChange: 'transform' },
+  } as const;
 
   return (
-    <Box as='div' {...outerProps} {...rest}>
-      <div {...innerProps} ref={innerProps.ref}>
-        {visibleChildren}
+    <Box
+      as='div'
+      style={{
+        minBlockSize: virtualizer.getTotalSize(),
+        minInlineSize: '100%',
+      }}
+      {...rest}
+    >
+      <div {...innerProps}>
+        {virtualizer.getVirtualItems().map((virtualItem) => (
+          <div
+            key={virtualItem.key}
+            data-index={virtualItem.index}
+            ref={virtualizer.measureElement}
+            style={{
+              position: 'absolute',
+              top: 0,
+              left: 0,
+              width: '100%',
+              transform: `translateY(${virtualItem.start}px)`,
```

Comment:
I feel like there should be a way to keep the negative margin we had before. Maybe you could adjust the `transform` (and the total size too)

The collapsing margins are important in the case of multi-select, to avoid showing multiple adjacent outlines. `Table` and `Tree` also have similar design decisions that would be nice to preserve when using virtualization.

Reply:
Okay, I will give that a try.

Reply:
So, updating the `transform` and total size didn't work. I think this is because the items are rendered based on the scroll distance and the reported element sizes, so when the list is scrolled to the bottom, the final few elements aren't rendered in. Video: 

https://github.com/iTwin/iTwinUI/assets/141063311/88e0793c-e7e5-49ba-8f2b-8a0a2e6f4afe

I did find a workaround for the spacing issue, though. A custom `measureElement` function can be passed into tanstack virtual [docs](https://tanstack.com/virtual/latest/docs/api/virtualizer#measureelement). This function allows us to treat the elements after the first in the list as shorter than they actually are. This impacts both the `translate` and total size, since they are based on the size results returned by `measureElement`, and avoids the scrolling issue, since the element sizes coming from the function now match the positioning we want them rendered at. Video: 

https://github.com/iTwin/iTwinUI/assets/141063311/9b509627-85ec-4a92-9bb6-c6d982c9e3e7

Added this function in this [commit](https://github.com/iTwin/iTwinUI/pull/2061/commits/6e7911a0cffc3a8b1b60213f41b8c4d2c9debd62).

Reply:
Did you try using a negative [`gap`](https://tanstack.com/virtual/latest/docs/api/virtualizer#gap) instead of manually measuring it?

Reply:
Yeah, using a gap of -1 gives slightly different spacing than using the custom `measureElement`. The custom `measureElement` is more accurate to what we originally have. Gap spacing: 
![image](https://github.com/iTwin/iTwinUI/assets/141063311/08c940c3-7a4f-4ddf-a445-948d00897cfc)
Measure Element spacing: 
![image](https://github.com/iTwin/iTwinUI/assets/141063311/200f1249-cb10-40dc-a8f2-5546d2e92140)
I'm not completely sure why these are different, but the second option is a pixel closer when using gap, which throws off the spacing by a pixel for all of the options. The overall difference is minor, so it might be better to use gap instead for readability.